### PR TITLE
CDCL engine

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libsolutil/CDCL.cpp [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+			m_clauses.emplace_back(make_unique<Clause>(move(learntClause)));
+			setupWatches(*m_clauses.back());
+		}
+		else
+		{
+			if (auto variable = nextDecisionVariable())
+			{
+				m_decisionPoints.emplace_back(m_assignmentTrail.size());
+				cout << "Deciding on " << m_variables.at(*variable) << " @" << currentDecisionLevel() << endl;
+				enqueue(Literal{false, *variable}, nullptr);
```
Comment (msooseth) [inline libsolutil/CDCL.cpp:84]: TODO: we should do polarity caching -- decide on the polarity of the variable it was last set. This is 2-3 lines to do and should bring considerable benefits. (here, we set the polarity fixed, with `Literal{false, *variable}`)

See: https://link.springer.com/chapter/10.1007/978-3-540-72788-0_28

File: libsolutil/CDCL.cpp [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+		{
+			if (currentDecisionLevel() == 0)
+			{
+				cout << "Insatisfiable" << endl;
+				return nullopt;
+			}
+			auto&& [learntClause, backtrackLevel] = analyze(move(*conflictClause));
+			cancelUntil(backtrackLevel);
+			m_clauses.emplace_back(make_unique<Clause>(move(learntClause)));
+			setupWatches(*m_clauses.back());
```
Comment (msooseth) [inline libsolutil/CDCL.cpp:90]: The correct way to do this is to `cancelUntil(backtrackLevel);`, attach the learnt clause to the watchlist, optinally add the learnt clause to the solver, and then enqueue the literal propagated by the newly learnt clause, with the `reason` set to the clause. Then call `propagate()`. I don't see the `enqueue` here.

File: test/libsolutil/CDCL.cpp [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+BOOST_AUTO_TEST_CASE(basic)
+{
+	auto x = variable("x");
+	Clause c1{x, ~x};
+	satisfiable({c1});
+}
+
+BOOST_AUTO_TEST_CASE(basic_unsat1)
+{
+	auto x = variable("x");
```
Comment (stackenbotten) [inline test/libsolutil/CDCL.cpp:69]: Coding style error

Comment (stackenbotten) [inline test/libsolutil/CDCL.cpp:69]: Coding style error

File: test/libsolutil/CDCL.cpp [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	auto x1 = variable("x1");
+	auto x2 = variable("x2");
+	Clause c1{x1, ~x2};
+	Clause c2{~x1, x2};
+	Clause c3{x1, x2};
+	Clause c4{~x1, ~x2};
+	unsatisfiable({c1, c2, c3, c4});
+}
+
+BOOST_AUTO_TEST_CASE(basic_sat)
```
Comment (msooseth) [inline test/libsolutil/CDCL.cpp:84]: Sorry, I named this wrong, it's supposed to be `basic_unsat3`

**Comment by cameel** [comment]
> Those are being rewritten in Rust anyway and are not part of the roadmap, so my guess would be close

Closing.

**Pull request closed without merging**